lib/logstash/pluginmanager/util.rb

```diff
@@ -1,64 +1,101 @@
+  # @param spec [Gem::Specification] plugin gem specification
+  # @return [Boolean] true if this spec is for an installable logstash plugin
+  def self.is_logstash_plugin_gem_spec?(spec)
+    spec.metadata["logstash_plugin"] == "true"
   end
 
-  def self.download_gem(gem_name, gem_version = '')
- 
-    gem_version ||= Gem::Requirement.default
- 
-    dep = ::Gem::Dependency.new(gem_name, gem_version)
-    specs_and_sources, errors = ::Gem::SpecFetcher.fetcher.spec_for_dependency dep
-    if specs_and_sources.empty?
-      return false
-    end
-    spec, source = specs_and_sources.max_by { |s,| s.version }
-    path = source.download( spec, java.lang.System.getProperty("java.io.tmpdir"))
-    path
+  # @param path [String] path to .gem file
+  # @return [Gem::Specification] .get file gem specification
+  # @raise [Exception] Gem::Package::FormatError will be raised on invalid .gem file format, might be other exceptions too
+  def self.plugin_file_spec(path)
+    Gem::Package.new(path).spec
   end
 
-  def self.installed?(name)
-    Gem::Specification.any? { |x| x.name == name }
+  # @param plugin [String] the plugin name or the local path to a .gem file
+  # @return [Boolean] true if the plugin is a local .gem file
+  def self.is_plugin_file?(plugin)
+    (plugin =~ /\.gem$/ && File.file?(plugin))
   end
 
-  def self.matching_specs(name)
-    req = Gem::Requirement.default
-    re = name ? /#{name}/i : //
-    specs = Gem::Specification.find_all{|spec| spec.name =~ re && req =~ spec.version}
-    specs.inject({}){|result, spec| result[spec.name_tuple] = spec; result}.values
+  # retrieve gem specs for all or specified name valid logstash plugins locally installed
+  # @param name [String] specific plugin name to find or nil for all plungins
+  # @return [Array<Gem::Specification>] all local logstash plugin gem specs
+  def self.find_plugins_gem_specs(name = nil)
+    specs = name ? Gem::Specification.find_all_by_name(name) : Gem::Specification.find_all
+    specs.select{|spec| is_logstash_plugin_gem_spec?(spec)}
   end
 
-  def self.ask_yesno(prompt)
-    while true
-      $stderr.puts ("#{prompt} [y/n]: ")
-      case $stdin.getc.downcase
-        when 'Y', 'y', 'j', 'J', 'yes' #j for Germans (Ja)
-          return true
-        when /\A[nN]o?\Z/ #n or no
-          break
-      end
-    end
+  # list of all locally installed plugins specs specified in the Gemfile.
+  # note that an installed plugin dependecies like codecs will not be listed, only those
+  # specifically listed in the Gemfile.
+  # @param gemfile [LogStash::Gemfile] the gemfile to validate against
+  # @return [Array<Gem::Specification>] list of plugin names
+  def self.all_installed_plugins_gem_specs(gemfile)
+    # we start form the installed gemspecs so we can verify the metadata for valid logstash plugin
+    # then filter out those not included in the Gemfile
+    find_plugins_gem_specs.select{|spec| !!gemfile.find(spec.name)}
```

Comment:
Do you really need to `!!` here? I think Enumerable/Array#select do fine with truthy results:

```
>> [1,2,3].select { |x| "fruit" }
=> [1, 2, 3]
```